File: backend/app/schema/lens.py
from typing import List

from pydantic import BaseModel

from .base import Completion

class Token(BaseModel):
    idx: int
    target_id: int

##### TARGETED LENS REQUEST SCHEMA #####

class TargetedLensCompletion(Completion):
    name: str
    tokens: List[Token]

class TargetedLensRequest(BaseModel):
    completions: List[TargetedLensCompletion]

##### TARGETED LENS RESPONSE SCHEMA #####

class Point(BaseModel):
    name: str
    prob: float

class LayerResults(BaseModel):
    layer: int
    points: List[Point]

class LensMetadata(BaseModel):
    maxLayer: int

class LensResponse(BaseModel):
    data: List[LayerResults]
    metadata: LensMetadata

##### GRID LENS REQUEST SCHEMA #####

class GridLensRequest(BaseModel):
    completion: Completion

##### GRID LENS RESPONSE SCHEMA #####

class GridLensResponse(BaseModel):
    id: str
    probs: List[List[float]]
    pred_strs: List[List[str]]

File: interface/src/types/lens.ts
import { Completion, Annotation, ChartMode } from "@/types/workspace"
import { LineGraphData } from "@/types/charts";
import { LensHeatmap, LensLineGraph } from "@/components/charts/types";

// Request Schema

export interface TokenCompletion {
    idx: number;
    target_id: number;
    target_text?: string;
}

export interface LensCompletion extends Completion { 
    name: string;
    model: string;
    tokens: TokenCompletion[];
}

export interface LineGraphAnnotation extends Annotation { 
    lineId: string;
    layer: number;
}

export interface LineGraphRangeAnnotation extends Annotation { 
    lineId: string;
    start: number;
    end: number;
}

export interface CellPosition {
    row: number;
    col: number;
}


export interface HeatmapAnnotation extends Annotation { 
    positions: CellPosition[];
}

export interface LogitLensWorkspace { 
    id?: string;
    name: string;
    completions: LensCompletion[];
    graphData: LineGraphData | null;
    annotations: LineGraphAnnotation[];
}

// Line Chart Data Schema

interface Point {
    name: string;
    prob: number;
}

interface Layer {
    layer: number;
    points: Point[];
}

export interface LensLineResponse {
    data: Layer[];
    metadata: {
        maxLayer: number;
    };
}

// Grid Chart Data Schema

export interface LensGridResponse {
    layer: number;
    probs: number[][];
    pred_strs: string[][];
}

// Chart Modes

export const LogitLensModes: ChartMode[] = [
    {
        name: "Target Token",
        description: "Probability of the target token per layer.",
        icon: "chart-area",
        component: LensLineGraph,
    },
    {
        name: "Prediction Grid",
        description: "Grid of the most probable token per layer.",
        icon: "grid-3x3",
        component: LensHeatmap,
    },
]
